Move soc code under ZEPHYR_PROJECT/soc

nashif opened the pull request:
Move the SoC outside of the architecture tree and put them at the same
level as boards and architectures allowing both SoCs and boards to be
maintained outside the tree.

nashif committed:
soc: support out of dir soc code

Just like boards, support also SoCs out of the tree.

Signed-off-by: Anas Nashif <anas.nashif@intel.com>

nashif committed:
x86: move soc/cpu to top-level under soc/

Move the SoC outside of the architecture tree and put them at the same
level as boards and architectures allowing both SoCs and boards to be
maintained outside the tree.

Signed-off-by: Anas Nashif <anas.nashif@intel.com>

nashif committed:
soc: adapt Kconfig/CMake for soc in top /

Change build system to point to soc/ in the top directory rather than
under arch/

Signed-off-by: Anas Nashif <anas.nashif@intel.com>

nashif committed:
kconfig: adapt help messages for new location of SoC code

Point to the new location of the soc code.

Signed-off-by: Anas Nashif <anas.nashif@intel.com>

nashif committed:
arm: soc: move arm SoCs to top-dir

Move the SoC outside of the architecture tree and put them at the same
level as boards and architectures allowing both SoCs and boards to be
maintained outside the tree.

Signed-off-by: Anas Nashif <anas.nashif@intel.com>

nashif committed:
arc: move soc to top-level dir soc/

Move the SoC outside of the architecture tree and put them at the same
level as boards and architectures allowing both SoCs and boards to be
maintained outside the tree.

Signed-off-by: Anas Nashif <anas.nashif@intel.com>

nashif committed:
xtensa: move soc to top-level dir soc/

Move the SoC outside of the architecture tree and put them at the same
level as boards and architectures allowing both SoCs and boards to be
maintained outside the tree.

Signed-off-by: Anas Nashif <anas.nashif@intel.com>

nashif committed:
riscv32: move soc to top-level dir soc/

Move the SoC outside of the architecture tree and put them at the same
level as boards and architectures allowing both SoCs and boards to be
maintained outside the tree.

Signed-off-by: Anas Nashif <anas.nashif@intel.com>

nashif committed:
nios2: move soc to top-level dir soc/

Move the SoC outside of the architecture tree and put them at the same
level as boards and architectures allowing both SoCs and boards to be
maintained outside the tree.

Signed-off-by: Anas Nashif <anas.nashif@intel.com>

nashif committed:
posix: move soc to top-level dir soc/

Move the SoC outside of the architecture tree and put them at the same
level as boards and architectures allowing both SoCs and boards to be
maintained outside the tree.

Signed-off-by: Anas Nashif <anas.nashif@intel.com>

nashif committed:
CODEOWNERS: adapt to new location of soc code

soc code is now under ZEPHYR_BASE/soc

Signed-off-by: Anas Nashif <anas.nashif@intel.com>

aescolar commented on CODEOWNERS:
```diff
@@ -6,22 +6,22 @@
+soc/arm/ti_simplelink/cc32xx             @GAnthony
+soc/arm/ti_simplelink/msp432p4xx         @Mani-Sadhasivam
 arch/nios2/                              @andrewboie @ramakrishnapallala
 arch/posix/                              @aescolar
```
An entry for soc/posix/ at my name would be logical then.
I guess similarly each other soc/{arch}/ would need a dedicated line with the maintainers in arch/{arch}

aescolar reviewed (commented):
Just a minor thing

nashif committed:
kconfig: move soc selection to ZEPHYR_BASE/soc/

Rather than do that for each architecture, source SoC Kconfigs where the
code is maintained, under ZEPHYR_BASE/soc.

Signed-off-by: Anas Nashif <anas.nashif@intel.com>

ulfalizer commented on soc/Kconfig:
```diff
@@ -0,0 +1,9 @@
+
+choice
+	prompt "SoC/CPU/Configuration Selection"
+	source "soc/$ENV_VAR_ARCH/*/Kconfig.soc"
```
Small subjective nit here. Could make it:

```
choice
	prompt "SoC/CPU/Configuration Selection"

source "soc/$ENV_VAR_ARCH/*/Kconfig.soc"

endchoice
```

Avoids confusing the `source` with a property on the `choice`.

ulfalizer replied:
(Unrelated: I wonder why it's `ENV_VAR_ARCH` instead of just `ARCH`.)

ulfalizer commented on soc/Kconfig:
```diff
@@ -0,0 +1,9 @@
+
+menu "Hardware Configuration"
+	source "soc/$ENV_VAR_ARCH/*/Kconfig"
+endmenu
```
Could maybe have

```
menu "Hardware Configuration"

source "soc/$ENV_VAR_ARCH/*/Kconfig"

endmenu
```

here too, just for consistency.

ulfalizer reviewed (commented):
`kconfig: move soc selection to ZEPHYR_BASE/soc/` looks OK. Seems like files are still getting included in the same order, which is the important thing.

dbkinder reviewed (approved):
doc changes look OK.

andyross reviewed (approved):
Didn't come close to viewing every line, but I agree with the architecture for sure.

nashif committed:
build: support out of tree SoC definition

Add the glue to make this work with SoC code outside of the tree.

Signed-off-by: Anas Nashif <anas.nashif@intel.com>

carlescufi commented on CMakeLists.txt:
```diff
@@ -50,9 +50,9 @@ zephyr_library_named(zephyr)
-  soc/${ARCH}/${SOC_PATH}
-  soc/${ARCH}/${SOC_PATH}/include
-  soc/${ARCH}/${SOC_FAMILY}/include
+  ${SOC_DIR}/${ARCH}/${SOC_PATH}
```
I wonder if we need a `SOC_DIR`. We already have a `BOARD_DIR`, so why not have instead a `ZEPHYR_OOT_DIR` that points to a folder that inside has `soc/` and `boards/`?

aescolar replied:
@carlescufi The SOC and the board may be in separate git repositories and placed in separate paths.

carlescufi replied:
I know, but is it really important that we allow this? Wouldn't it be better to allow out of tree boards and SoCs, but keeping the same folder structure as in the Zephyr repo itself? Otherwise we might end up with a whole bunch of env vars like those?

aescolar replied:
I guess it is not too important. It just adds a bit more of flexibility/convenience, but nothing one cannot work relatively easy around.

> Otherwise we might end up with a whole bunch of env vars like those?
An ARCH_DIR would be the next logical one I'd be happy to see :)

nashif replied:
> I wonder if we need a SOC_DIR. We already have a BOARD_DIR, so why not have instead a ZEPHYR_OOT_DIR that points to a folder that inside has soc/ and boards/?

the use case I am implementing right now is soc and board codes in different git repos. This was also tested with both in the same structure resembling the zephyr tree. (both variables point to the same thing)

I do not think having another variable is excessive here (compare with the 2839485 Kconfig variables we have :-)

carlescufi replied:
Fine, if you rely on that then no objections.

aescolar commented on CODEOWNERS:
```diff
@@ -6,23 +6,27 @@
+soc/arm/ti_simplelink/cc32xx             @GAnthony
+soc/arm/ti_simplelink/msp432p4xx         @Mani-Sadhasivam
 arch/nios2/                              @andrewboie @ramakrishnapallala
 arch/posix/                              @aescolar
```
soc/posix/ is orphan now. Please add a line assigned to me

nashif committed:
doc: custom SOC configurations

Document how to use out-of-tree SOC configurations.

Signed-off-by: Anas Nashif <anas.nashif@intel.com>
